support for dotnet based projects

durga-pasupuleti opened the pull request.

prabhu commented:
@dplabsindia, Thank you so much. Could you add some repotests as well?

prabhu commented on index.js:
```diff
@@ -4574,6 +4574,9 @@ export async function createCsharpBom(path, options) {
     !paketLockFiles.length
   ) {
     const filesToRestore = slnFiles.concat(csProjFiles);
+    console.error(
```
I think this message would always appear even before we restore?

prabhu commented on index.js:
```diff
@@ -4702,9 +4706,44 @@ export async function createCsharpBom(path, options) {
+        
+        if(dependsOn.name && dependsOn.version){ 
+          //console.log("prefixLLLL: ",prefix);
+          let dependcy = prefix+"/";
```
Use template [strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals).

prabhu commented on index.js:
```diff
@@ -4726,28 +4765,34 @@ export async function createCsharpBom(path, options) {
       if (DEBUG_MODE) {
         console.log(`Parsing ${f}`);
       }
+      console.log(`csProjFiles List Size :: ${csProjFiles.length}`);
```
These extra console.logs could be removed or wrapped behind `if (DEBUG_MODE)`

prabhu commented on index.js:
```diff
@@ -4771,7 +4816,8 @@ export async function createCsharpBom(path, options) {
     const retMap = await getNugetMetadata(pkgList, dependencies);
-    if (retMap.dependencies?.length) {
+    console.log(`dependencies re${retMap.dependencies?.length}`);
+    if (retMap.dependencies?.length > 1) {
```
Why greater than 1?

durga-pasupuleti replied:
There is an empty set for this. It adds nothing.

prabhu commented on index.js:
```diff
@@ -4884,11 +4930,13 @@ export function mergeDependencies(
         provides: Array.from(provides_map[akey]).sort(),
       });
     } else {
+       if (deps_map && deps_map.size > 1){        
```
Why > 1 and not 0?

durga-pasupuleti replied:
Empty set not needed for the sbom.

prabhu replied:
> 1 means it will skip objects with just 1 entry.

prabhu commented:
- [ ] Run `npm run lint` command to resolve biome issues
- [ ] Needs unit tests and a repo tests for .Net framework

prabhu commented:
@durga-pasupuleti Are you planning to implement the PR comments, so that we can include this in 10.6.x?

durga-pasupuleti commented:
Yes Prabhu. Today i will push them

On Mon, May 20, 2024 at 14:40 prabhu ***@***.***> wrote:

> @durga-pasupuleti <https://github.com/durga-pasupuleti> Are you planning
> to implement the PR comments, so that we can include this in 10.6.x?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CycloneDX/cdxgen/pull/1088#issuecomment-2121083444>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFK723RD2BRXUZSNQI6X4HTZDJGTTAVCNFSM6AAAAABH3F5NZKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRRGA4DGNBUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

prabhu commented on index.js:
```diff
@@ -4702,9 +4703,38 @@ export async function createCsharpBom(path, options) {
+      parentDependsOn.forEach((dependsOn) => {
+        if(dependsOn.name && dependsOn.version){ 
+          //console.log("prefix: ",prefix);
+          let dependcy = `${prefix}/`;
```
All the below lines can be combined to using template strings. This way we don't have to append strings together.

prabhu commented on index.js:
```diff
@@ -4772,10 +4769,11 @@ export async function createCsharpBom(path, options) {
       }
-
-      console.log(`pkgList List Size :: ${pkgList.length}`);
+      if(DEBUG_MODE){
```
Please run `npm run lint` which will fix the linting issues automatically. Thank you so much for your patience.

durga-pasupuleti replied:
I ran the lint. For some reason on my machine its producing error for directory name.

prabhu replied:
Could you share the error?

@setchy any ideas?

durga-pasupuleti replied:
Os error 123, the filename , directory name or voe label syntax is incorrect

prabhu replied:
That's a strange error. Could you contribute a test repo to add to the repotests? I will take care of the linting errors.

prabhu commented:
Could you kindly follow the instructions mentioned here to sign your commits?

https://github.com/CycloneDX/cdxgen/pull/1088/checks?check_run_id=25230419605

I think we are very close to getting this tested and merged!

prabhu commented on index.js:
```diff
@@ -4884,10 +4916,12 @@ export function mergeDependencies(
-        ref: akey,
-        dependsOn: Array.from(deps_map[akey]).sort(),
-      });
+       if (deps_map && deps_map.size > 0){        
```
Is there a method called size? I usually do Object.keys(deps_map).length

prabhu commented on index.js:
```diff
@@ -4702,9 +4703,35 @@ export async function createCsharpBom(path, options) {
+    }
+    if (parentDependsOn.size) {
+      const arrrr =  new Set();
+      const prefix = parentComponent["bom-ref"].split("/")[0];
```
If prefix is just `pkg:nuget` we can directly used that in 4723.

durga-pasupuleti replied:
For some reason this without this its not able to create the dependencies. I checked multiple times for this.

prabhu commented on index.js:
```diff
@@ -4702,9 +4703,35 @@ export async function createCsharpBom(path, options) {
+
+      if (dlist?.length) {
+        for (const p of dlist) {
+          parentDependsOn.add(p);
```
Can this not store the bom-ref similar to 4723. I think we don't need another loop in 4720.

durga-pasupuleti replied:
Same as above, with these changes only i am able to generate sbom with dependencies.

prabhu replied:
The second loop appears unnecessary. It is only checking if `p` has both name and version.

durga-pasupuleti replied:
Yes you are correct, Its a overlook. I just updated the code.

prabhu commented on index.js:
```diff
@@ -4873,12 +4894,16 @@ export async function createCsharpBom(path, options) {
-        `Found ${pkgList.length} components by parsing the ${csProjFiles.length} csproj files. The resulting SBOM will be incomplete.`,
-      );
-      options.failOnError && process.exit(1);
+     // console.log(
```
Any ideas why this is commented out? This message is useful to indicate the absence of lock files for all users. Perhaps, we can suppress this only for .Net framework somehow.

durga-pasupuleti replied:
Yeah we should find a way for saying the incomplete sbom. I dont think its best idea based on this.

durga-pasupuleti commented:
> Could you kindly follow the instructions mentioned here to sign your commits?
> 
> 
> 
> https://github.com/CycloneDX/cdxgen/pull/1088/checks?check_run_id=25230419605
> 
> 
> 
> I think we are very close to getting this tested and merged!

I have added signoff by following the steps and added repotests. Anything i can do from my side.

prabhu commented:
@durga-pasupuleti, Thank you for this PR. It gave me the idea about what is missing. Let me collect some more examples and reimplement this slightly differently.

prabhu commented:
Superseded by #1119. Thank you so much for your guidance.
